File: r/factset.protobuf.stach/R/NullValues.R
#' @docType class
#' @title factset.protobuf.stach.NullValues
#' @description Class defining the NullValues of various data types
#' @format An \code{R6Class} generator object
#' @field DOUBLE Null value for DOUBLE data type
#' @field FLOAT Null value for FLOAT data type
#' @field INT32 Null value for INT32 data type
#' @field INT64 Null value for INT64 data type
#' @field STRING Null value for STRING data type
#' @field TIMESTAMP Null value for TIMESTAMP data type
#' @field DURATION Null value for DURATION data type
#' @examples
#' nullValues <- factset.protobuf.stach.NullValues$new()
#'
#' # Null value comparison for Protobuf's String data type.
#' a <- 'null'
#' if(nullValues$STRING == a) print("a is the null") else print("a is not a null")
#'
#' # Null value comparison for Protobuf's Int32 data type.
#' b <- new(google.protobuf.Int32Value, value = 2147483647)
#' if(nullValues$INT32 == b$value) print("b is the null") else print("b is not a null")
#'
#' # Null value comparison for Protobuf's Duration data type.
#' c <- new(google.protobuf.Duration, seconds = 315576000000, nanos = 999999900)
#' if(nullValues$DURATION == c) print("c is the null") else print("c is not a null")
#' @importFrom R6 R6Class
#' @export
factset.protobuf.stach.NullValues <- R6::R6Class(
  'factset.protobuf.stach.NullValues',
  public = list(

    DOUBLE = NaN,
    FLOAT = NaN,
    INT32 = 2147483647,
    INT64 = 9223372036854775807,
    STRING = "null",
    TIMESTAMP = NULL,
    DURATION = NULL,

    initialize = function(){
      library(RProtoBuf)
      self$TIMESTAMP = new(google.protobuf.Timestamp, seconds = 253402300799, nanos = 0)
      self$DURATION = new(google.protobuf.Duration, seconds = 315576000000, nanos = 999999900)
    }
  )
)

File: js/null-values.js
"use strict";
var stach = require('./stach');
var Long = require('long');

const NullValues = {
  DOUBLE: Number.NaN,
  FLOAT: Number.NaN,
  INT32: 2147483647,
  INT64: Long.fromString('9223372036854775807'),
  STRING: 'null',
  DURATION: new stach.google.protobuf.Duration({
    seconds: Long.fromString('315576000000'),
    nanos: 999999900
  }),
  TIMESTAMP: new stach.google.protobuf.Timestamp({
    seconds: Long.fromString('253402300799'),
    nanos: 0
  }),

  isNull(value, dataType) {
    switch (dataType) {
      case stach.factset.protobuf.stach.table.DataType.DOUBLE:
      case stach.factset.protobuf.stach.table.DataType.FLOAT:
        return Number.isNaN(value);

      case stach.factset.protobuf.stach.table.DataType.INT32:
        return NullValues.INT32 === value;

      case stach.factset.protobuf.stach.table.DataType.INT64:
        return NullValues.INT64.equals(value);

      case stach.factset.protobuf.stach.table.DataType.STRING:
        return NullValues.STRING === value;

      case stach.factset.protobuf.stach.table.DataType.DURATION:
        return value instanceof stach.google.protobuf.Duration && NullValues.DURATION.seconds.equals(value.seconds) && NullValues.DURATION.nanos === value.nanos;

      case stach.factset.protobuf.stach.table.DataType.TIMESTAMP:
        return value instanceof stach.google.protobuf.Timestamp && NullValues.TIMESTAMP.seconds.equals(value.seconds) && NullValues.TIMESTAMP.nanos === value.nanos;

      default:
        return false;
    }
  }

};
stach.factset.protobuf.stach.NullValues = NullValues;
